magma/circuit.py

```diff
@@ -107,7 +111,8 @@ def _maybe_add_default_clock(cls):
 
     # Check all instances in @cls for an undriven clock port.
     inst_ports = itertools.chain(
-        *(instance.interface.ports.values() for instance in cls.instances))
+        *(instance.interface.ports.values() for instance in cls.instances
+          if not getattr(instance, "combinational", False)))
```

Comment:
Yes, I added it when I was trying to optimize the Concat primitive version.  I updated it to use the attribute `"stateful"` instead of `combinational` since that's an existing interface (used by the Python simulator to determine stateful primitives).  Basically, if the user says `stateful = False` we can skip doing the automatic clock wiring traversal logic.  Not as important now since we're not using the concat tree pattern, but figured it's a nice bonus to keep around

Reply:
A related improvement for the general case could be, when we finalize an interface, we can check if there are any clock ports, and mark the definition as stateful=False, so when it's instanced, the automatic clock wiring logic for the container can skip traversing an instance.  Related, if we're already doing this traversal at defn time, we could populate a list of clock ports rather than having to generically traverse all instance ports.